paddle/fluid/pybind/reader_py.cc

```diff
@@ -20,46 +20,134 @@
 #include <utility>
 #include <vector>
 #include "Python.h"
+#include "boost/optional.hpp"
+#include "gflags/gflags.h"
 #include "paddle/fluid/framework/ddim.h"
 #include "paddle/fluid/framework/reader.h"
 #include "paddle/fluid/imperative/layer.h"
 #include "paddle/fluid/imperative/tracer.h"
 #include "paddle/fluid/operators/reader/buffered_reader.h"
+#include "paddle/fluid/operators/reader/lod_tensor_blocking_queue.h"
 #include "paddle/fluid/operators/reader/py_reader.h"
 #include "paddle/fluid/platform/place.h"
 #include "pybind11/stl.h"
 
+DEFINE_bool(reader_queue_speed_test_mode, false,
+            "If set true, the queue.pop will only get data from queue but not "
+            "remove the data from queue for speed testing");
+
 namespace paddle {
 namespace pybind {
 
 namespace py = pybind11;
+namespace reader = operators::reader;
+
+// Check whether the tensor shape matches the VarDesc shape
+// Return the different shape if exists
+static boost::optional<std::vector<int64_t>> DiffTensorShapeWithVarDesc(
+    const framework::LoDTensor &tensor, const framework::VarDesc &var_desc,
+    size_t num_places) {
+  auto tensor_shape = tensor.dims();
+  auto desc_shape = var_desc.GetShape();
+
+  int64_t rank = tensor_shape.size();
+
+  if (UNLIKELY(rank == 0)) {
+    if (desc_shape.size() != 0) {  // Tensor rank = 0 but desc does not match
+      return framework::vectorize<int64_t>(tensor_shape);
+    } else {
+      return boost::none;
+    }
+  }
+
+  PADDLE_ENFORCE_GE(tensor_shape[0], 0,
```

Comment:
need to explain here is the 0th dimension of shape?

Reply:
Added, thanks.